From 7e2663f3244af478c200fc51994fbda85fe54476 Mon Sep 17 00:00:00 2001
From: xianbing Zhu <xianbing.zhu@linux.alibaba.com>
Date: Thu, 30 May 2024 11:45:17 +0800
Subject: [PATCH 062/129] mmc:sdhci-of-dwcmshc: th1520 resolve accss rpmb error
 in hs400

After complete selection of HS400 ,software reset DAT & cmd line
to resolve first time data access  error(time out) when
first swith to hs400 mode,both in probe emmc and switch to
access rpmb.

Signed-off-by: xianbing Zhu <xianbing.zhu@linux.alibaba.com>
---
 drivers/mmc/host/sdhci-of-dwcmshc.c | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

--- a/drivers/mmc/host/sdhci-of-dwcmshc.c
+++ b/drivers/mmc/host/sdhci-of-dwcmshc.c
@@ -642,6 +642,29 @@ static void th1520_sdhci_reset(struct sd
 	}
 }
 
+/* Complete selection of HS400 ,software reset DAT & cmd line
+* resolve for first time data access  error(time out) when
+* first swith to hs400 mode.
+*
+* Some (ENE) controllers go apeshit on some ios operation,
+* signalling timeout and CRC errors even on CMD0. Resetting
+* it on each ios seems to solve the problem.
+*
+*/
+static void  th1520_sdhci_hs400_complete(struct mmc_host *mmc)
+{
+	struct sdhci_host *host = mmc_priv(mmc);
+	u8 mask = SDHCI_RESET_CMD | SDHCI_RESET_DATA;
+
+	if (host->quirks & SDHCI_QUIRK_NO_CARD_NO_RESET) {
+
+		if (!mmc->ops->get_cd(mmc))
+			return;
+	}
+	th1520_sdhci_reset(host,mask);
+
+}
+
 static const struct sdhci_ops sdhci_dwcmshc_ops = {
 	.set_clock		= sdhci_set_clock,
 	.set_bus_width		= sdhci_set_bus_width,
@@ -889,6 +912,8 @@ static int dwcmshc_probe(struct platform
 		}
 
 		sdhci_enable_v4_mode(host);
+
+		host->mmc_host_ops.hs400_complete = th1520_sdhci_hs400_complete;
 	}
 
 #ifdef CONFIG_ACPI
